XenAdmin/Diagnostics/Checks/VSwitchControllerCheck.cs

```diff
@@ -0,0 +1,91 @@
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Collections.Generic;
+using XenAdmin.Core;
+using XenAdmin.Diagnostics.Problems;
+using XenAdmin.Diagnostics.Problems.PoolProblem;
+using XenAPI;
+
+namespace XenAdmin.Diagnostics.Checks
+{
+    class VSwitchControllerCheck : HostPostLivenessCheck
+    {
+        private readonly Dictionary<string, string> _installMethodConfig;
+        private readonly Pool _pool;
+        private XenServerVersion _newVersion;
+
+        public VSwitchControllerCheck(Host host, XenServerVersion newVersion)
+            : base(host)
+        {
+            _newVersion = newVersion;
+            _pool = Helpers.GetPoolOfOne(Host?.Connection);
+        }
+
+        public VSwitchControllerCheck(Host host, Dictionary<string, string> installMethodConfig)
+            : base(host)
+        {
+            _installMethodConfig = installMethodConfig;
+            _pool = Helpers.GetPoolOfOne(Host?.Connection);
+        }
+
+        public override string Description => Messages.CHECKING_VSWITCH_CONTROLLER;
+
+        protected override Problem RunHostCheck()
+        {
+            if (!_pool.vSwitchController() || Helpers.StockholmOrGreater(Host))
+                return null;
+
+            //update case
+            if (_newVersion != null)
+            {
+                if (_newVersion.Version.CompareTo(new Version(BrandManager.ProductVersion82)) >= 0)
+                    return new VSwitchControllerProblem(this, _pool);
+                return null;
+            }
+
+            //upgrade case
+            string upgradePlatformVersion = null;
+
+            if (_installMethodConfig != null)
+                Host.TryGetUpgradeVersion(Host, _installMethodConfig, out upgradePlatformVersion, out _);
+                
+            // we know they are upgrading to Stockholm or greater, so block them
+            if (Helpers.StockholmOrGreater(upgradePlatformVersion))
+                return new VSwitchControllerProblem(this, _pool);
+
+            // we don't know the upgrade version, so add generic warning

```

Comment:
If we know the platform version (upgradePlatformVersion is not null) and it is not Stockholm or greater, then we shouldn't add the warning (for example upgrading to 8.1)